packages/mcdoc/src/runtime/checker/index.ts

```diff
@@ -994,45 +1003,55 @@ function simplifyIndexed<T>(
 }
 
 function simplifyUnion<T>(typeDef: UnionType, context: SimplifyContext<T>): SimplifiedMcdocType {
-	const members: SimplifiedMcdocTypeNoUnion[] = []
 	const filterCanonical = context.ctx.requireCanonical
 		&& typeDef.members.some(m => m.attributes?.some(a => a.name === 'canonical'))
-	for (const member of typeDef.members) {
-		if (filterCanonical && !member.attributes?.some(a => a.name === 'canonical')) {
-			continue
-		}
-		const simplified = simplify(member, context)
-		let keep = true
-		handleAttributes(member.attributes, context.ctx, (handler, config) => {
-			if (!keep || !handler.filterElement) return
-			if (!handler.filterElement(config, context.ctx)) {
-				keep = false
+	const flatMembers = getPossibleTypes(typeDef)
+	const validMembers = flatMembers
+		.filter(member => {
+			if (filterCanonical && !member.attributes?.some(a => a.name === 'canonical')) {
+				return false
 			}
+			let keep = true
+			handleAttributes(member.attributes, context.ctx, (handler, config) => {
+				if (!keep || !handler.filterElement) return
+				if (!handler.filterElement(config, context.ctx)) {
+					keep = false
+				}
+			})
+			return keep
 		})
-		if (!keep) continue
 
-		if (simplified.kind === 'union') {
-			members.push(...simplified.members)
-		} else {
-			members.push(simplified)
-		}
+	let memberContext = context
+	if (context.literalStructFields && validMembers.length > 1) {
+		memberContext = { ...context, spreadingUnion: true }
 	}
+	const members = validMembers.map(member => simplify(member, memberContext))
```

Comment:
Actually now that I think about it, this will not work 100% correctly when one of the types resolves to an empty union later (e.g. if there are two valid members, one of which was a reference which itself resolves to an empty union).

Reply:
Fixed, without meaningful slowdown
![image](https://github.com/SpyglassMC/Spyglass/assets/12124394/a24be41e-3f3f-4006-8685-31793a8d9067)